tpm/src/socket.rs

```diff
@@ -0,0 +1,160 @@
+        self.state = SocketDevState::Connecting;
+
+        let now = Instant::now();
+
+        // Retry connecting for 10 secs
+        loop {
+            match UnixStream::connect(socket_path) {
+                Ok(s) => {
+                    let fd = s.as_raw_fd();
+                    self.ctrl_fd = fd;
+                    self.stream = Some(s);
+                    self.state = SocketDevState::Connected;
+                    debug!("Connected to tpm socket path : {:?}", socket_path);
+                    return Ok(());
+                }
+                Err(_e) => {}
+            };
+            sleep(Duration::from_millis(1000));
+
+            if now.elapsed().as_secs() >= 10 {
+                break;
+            }
+        }
+        Err(Error::ConnectToSocket(anyhow!(
+            "Failed to connect to tpm Socket"
+        )))
+    }
+
+    pub fn set_datafd(&mut self, fd: RawFd) {
+        self.data_fd = fd;
+    }
+
+    pub fn set_msgfd(&mut self, fd: RawFd) {
+        self.write_msgfd = fd;
+    }
+
+    pub fn send_full(&self, buf: &[u8]) -> Result<usize> {
+        let write_fd = self.write_msgfd;
+
+        let size = self
+            .stream
+            .as_ref()
+            .unwrap()
+            .send_with_fd(buf, write_fd)
+            .map_err(|e| {
+                Error::WriteToSocket(anyhow!("Failed to write to Socket. Error: {:?}", e))
+            })?;
+
+        Ok(size)
+    }
+
+    pub fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        if self.stream.is_none() {
+            return Err(Error::WriteToSocket(anyhow!(
+                "Stream for tpm socket was not initialized"
+            )));
+        }
+        let res = match self.state {
+            SocketDevState::Connected => {
+                let ret = self.send_full(buf)?;
+                // swtpm will receive data Fd after a successful send
+                // Reset cached write_msgfd after a successful send
+                // Ideally, write_msgfd is reset after first Ctrl Command
+                if ret > 0 && self.write_msgfd != 0 {
+                    self.write_msgfd = 0;
+                }
+                ret
+            }
+            _ => {
+                return Err(Error::WriteToSocket(anyhow!(
+                    "TPM Socket was not in Connected State"
+                )))
+            }
+        };
+        Ok(res)
+    }
+
+    pub fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let newbuf: &mut [u8] = &mut [0; TPM_CRB_BUFFER_MAX];
+
+        if self.stream.is_none() {
+            return Err(Error::ReadFromSocket(anyhow!(
+                "Stream for tpm socket was not initialized"
+            )));
+        }
+        let mut socket = self.stream.as_ref().unwrap();
+        let size: usize = socket.read(newbuf).map_err(|e| {
+            Error::ReadFromSocket(anyhow!("Failed to read from socket. Error Code {:?}", e))
+        })?;
+        buf[0..size].clone_from_slice(&newbuf[0..size]);
```

Comment:
If `buf` is larger than the bytes written this will succeed but how will the caller know that fewer bytes were read?

Reply:
CRB interface defines the max buf size to TPM_CRB_BUFFER_MAX(3968). I am using a buf of this size to read into. So, I am expecting there will be no case when I will only read a subset of bytes from backend. 

I have also verified all the Control commands implemented in this code. All the responses will fit in this MAX buff size. Hope this addresses your concern.